File: miaobian.py
# Python和C++的逻辑如出一辙，可以说是一模一样，没有繁琐的Mat定义
# 小窗imshow的时候不要用中文，其他和C++差不多，就不写那么多注释了
import cv2
import numpy as np
def main():
    # 读取原图片
    ma = cv2.imread('/home/chunshouy/桌面/ma.png')
    # 转化为灰度图片
    ma_gray = cv2.cvtColor(ma, cv2.COLOR_BGR2GRAY)
    # 转化为黑白图
    _, ma_bw = cv2.threshold(ma_gray, 127, 255, cv2.THRESH_BINARY_INV)

    # 膨胀 - 让文字区域变大
    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (15, 15))
    ma_bigger = cv2.dilate(ma_bw, kernel)

    # 找轮廓
    # contours：找到的轮廓列表每个轮廓是一个点的数组
    # _：层次信息（这里用不到，所以用下划线忽略）
    contours, _ = cv2.findContours(ma_bigger, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    # 在原图上画红色轮廓
    #先复制一下原图
    redoutline_ma = ma.copy()
    cv2.drawContours(redoutline_ma, contours, -1, (0, 0, 255), 3)

    # 保存图片
    cv2.imwrite('/home/chunshouy/桌面/outline_ma.jpg', redoutline_ma)

    # 显示结果
    cv2.imshow('redoutline', redoutline_ma)
    cv2.waitKey(0)
if __name__ == "__main__":
    main()

File: miaobian.cpp
//灰度图--二值化--膨胀--描边
#include <iostream>
#include <opencv2/opencv.hpp>
#include <string>
using namespace std;
using namespace cv;

int main() {
    //定义原图片路径
    string ma_path = "/home/chunshouy/桌面/ma.png";
    //读取原图片
    Mat ma = imread(ma_path);

    //转化为灰度图片
    Mat ma_gray;
    cvtColor(ma, ma_gray, COLOR_BGR2GRAY);

    //转化为黑白图
    // threshold(输入图像，输出图像，阈值，最大值，THRESH_BINARY)阈值=128；大于128变白(255)，小于128变黑(0)
    Mat ma_bw;
    // 使用 THRESH_BINARY_INV 让文字变成白色，背景变成黑色
    threshold(ma_gray, ma_bw, 128, 255, THRESH_BINARY_INV);

    //膨胀
    Mat ma_bigger; //膨胀后的变量
    
    // getStructuringElement(): 创建结构元素，
    // MORPH_RECT是矩形结构，Size()规定出来方框大小
    // 使用更大的核尺寸来充分膨胀文字
    Mat kernel = getStructuringElement(MORPH_RECT, Size(15, 15));
    
    // dilate(原图，目标图，kernel)：膨胀函数，让白色区域变大，kernel是小矩阵
    dilate(ma_bw, ma_bigger, kernel);

    //找轮廓
    // vector<vector<Point>> contours：创建存储轮廓的向量
    // findContours(定义的膨胀后的变量)：查找轮廓函数
    // ma_bigger：输入的膨胀后图像
    // contours：输出的轮廓集合
    // RETR_EXTERNAL：只检测最外层轮廓
    // CHAIN_APPROX_SIMPLE：简化轮廓点
    // 这块好抽象
    vector<vector<Point>> contours;
    findContours(ma_bigger, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);

    //画红线
    Mat redoutline_ma; //定义红色边框的变量
    
    //复制原图
    redoutline_ma = ma.clone();
    
    //找到所有的轮廓像素
    //红色的BGR: Scalar(0, 0, 255)
    for (int i = 0; i < contours.size(); i++) {
        drawContours(redoutline_ma, contours, i, Scalar(0, 0, 255), 3);
    } //这里的3是线宽

    //保存图片
    string outLine_path = "/home/chunshouy/桌面/outline_ma.jpg";
    imwrite(outLine_path, redoutline_ma);
    
    //展示图片
    imshow("红色轮廓", redoutline_ma);
    imshow("原图",ma);
    waitKey(0);
    
    return 0;
}